### PR TITLE
sync: Added support for syncing notary/cosign signatures, closes #261

### LINKED ISSUE #261: sync: support sync-ing image signatures
At the moment zot sync does not support copy-ing image signatures.

### PR DESCRIPTION
**Pull request by peusebiu**
Signed-off-by: Petu Eusebiu <peusebiu@cisco.com>

**What type of PR is this?**
feature

**Which issue does this PR fix**:
#261 

**What does this PR do / Why do we need it**:
Added support for syncing notary and cosign signatures

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**: 

**Automation added to e2e**:

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
no

**Does this change require updates to the CNI daemonset config files to work?**:

**Does this PR introduce any user-facing change?**:

no
```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by peusebiu** [comment]
will be updated with tests, just wanted to see code coverage

File: pkg/extensions/sync/on_demand.go [hunk 1]
```diff
@@ -125,13 +132,28 @@ func OneImage(cfg Config, storeController storage.StoreController,
 		} else {
 			log.Info().Msgf("successfully synced %s", upstreamRef.DockerReference().Name())
 
-			err := pushSyncedLocalImage(repo, tag, uuid.String(), storeController, log)
-			if err != nil {
+			if err = retry.RetryIfNecessary(context.Background(), func() error {
```
Comment (rchincha) [inline pkg/extensions/sync/on_demand.go:135]: what is this doing?

File: pkg/extensions/sync/sync.go [hunk 2]
```diff
@@ -240,6 +200,11 @@ func imagesToCopyFromUpstream(registryName string, repos []string, upstreamCtx *
 		}
 
 		for _, tag := range tags {
+			// don't copy cosign signature, skopeo doesn't support it
```
Comment (rchincha) [inline pkg/extensions/sync/sync.go:203]: s/skopeo/"containers/image"/

go over the rest of the code and do the same

File: pkg/extensions/sync/utils.go [hunk 3]
```diff
@@ -89,6 +100,284 @@ func getFileCredentials(filepath string) (CredentialsFile, error) {
...
+	return client, nil
+}
+
+func syncCosignSignature(client *resty.Client, storeController storage.StoreController,
+	regURL url.URL, repo, digest string, log log.Logger) error {
+	log.Info().Msg("syncing cosign signatures")
+
+	getCosignManifestURL := regURL
+
+	cosignEncodedDigest := strings.Replace(digest, ":", "-", 1) + ".sig"
```
Comment (rchincha) [inline pkg/extensions/sync/utils.go:151]: does cosign have any utility libs/apis to do this for us?

File: pkg/extensions/sync/utils.go [hunk 4]
```diff
@@ -89,6 +100,284 @@ func getFileCredentials(filepath string) (CredentialsFile, error) {
...
+
+func syncNotarySignature(client *resty.Client, storeController storage.StoreController,
+	regURL url.URL, repo, digest string, log log.Logger) error {
+	log.Info().Msg("syncing notary signatures")
+
+	getReferrersURL := regURL
+	getRefManifestURL := regURL
+
+	// based on manifest digest get referrers
+	getReferrersURL.Path = path.Join(getReferrersURL.Path, "oras/artifacts/v1/", repo, "manifests", digest, "referrers")
```
Comment (rchincha) [inline pkg/extensions/sync/utils.go:249]: same, does notation lib have helper libs?

File: pkg/extensions/sync/on_demand.go (hunk 1 above)
Comment (peusebiu) [reply pkg/extensions/sync/on_demand.go:135]: will remove it, added it by mistake, thanks for spotting it.

File: pkg/extensions/sync/utils.go (hunk 3 above)
Comment (peusebiu) [reply pkg/extensions/sync/utils.go:151]: they have one but is not exported...

File: pkg/extensions/sync/utils.go (hunk 4 above)
Comment (peusebiu) [reply pkg/extensions/sync/utils.go:252]: no, they are doing the same

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**